src/mrnet/core/rates.py

```diff
@@ -424,11 +408,11 @@ def calculate_rate(
         k_rate = self.calculate_rate_constant(temperature=temperature, reverse=reverse)
 
         if reverse:
-            exponents = np.array(list(self.rate_law["products"].values()))
-            mols = [p.mol_graph.molecule for p in self.products]
+            exponents = self.rate_law["products"]
+            mols = self.pro_mols
         else:
-            exponents = np.array(list(self.rate_law["reactants"].values()))
-            mols = [r.mol_graph.molecule for r in self.reactants]
+            exponents = self.rate_law["reactants"]
+            mols = self.rct_mols
 
         masses = [m.composition.weight for m in mols]
```

Comment:
Ah. I see that this is a reason to keep the molecules - we need masses. Well, again, I'd say the molecules themselves aren't useful. If anything, then, just store the weights. Store only what you directly need.

Reply:
Addressed in recent commit.